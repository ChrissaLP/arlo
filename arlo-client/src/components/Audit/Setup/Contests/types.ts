export interface IChoiceValues {
  id?: string
  name: string
  numVotes: string | number
}

export interface IContestValues {
  name: string
  isTargeted: boolean
  totalBallotsCast: string
  numWinners: string
  votesAllowed: string
  choices: IChoiceValues[]
}

export interface IValues {
  contests: IContestValues[]
}
