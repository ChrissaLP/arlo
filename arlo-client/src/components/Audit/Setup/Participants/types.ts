export interface IValues {
  state: string
}
