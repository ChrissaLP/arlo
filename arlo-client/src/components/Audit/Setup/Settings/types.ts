import { IAuditSettings } from '../../../../types'

export interface IValues {
  electionName: IAuditSettings['electionName']
  online: IAuditSettings['online']
  randomSeed: IAuditSettings['randomSeed']
  riskLimit: IAuditSettings['riskLimit']
}
