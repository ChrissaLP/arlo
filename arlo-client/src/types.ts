export interface CreateAuditParams {
  electionId: string
}

export interface AuditFlowParams extends CreateAuditParams {
  token: string
  round?: string
  ballot?: string
}

export interface Candidate {
  id: string
  name: string
  numVotes: number | string
}

export interface SampleSizeOption {
  size: number | string
  prob: number | null
  type: string | null
}

export interface Contest {
  id: string
  name: string
  choices: Candidate[]
  totalBallotsCast: string
}

export interface AuditMember {
  name: string
  affiliation: 'DEM' | 'REP' | 'LIB' | 'IND' | ''
}

export interface Ballot {
  tabulator: string
  batch: string
  record: string
  status: 'AUDITED' | 'NOT_AUDITED'
  vote: 'YES' | 'NO' | 'NO_CONSENSUS' | 'NO_VOTE' | null
}

export interface AuditBoard {
  id: string
  name: string
  members: [AuditMember, AuditMember] | []
  ballots?: Ballot[]
}

export interface BallotManifest {
  filename: string | null
  numBallots: number | '' | null
  numBatches: number | '' | null
  uploadedAt: string | null
}

export interface Jurisdiction {
  id: string
  name: string
  contests: string[]
  auditBoards: AuditBoard[]
  ballotManifest?: BallotManifest
}

export interface RoundContest {
  id: string
  results: {
    [key: string]: number
  }
  sampleSizeOptions: SampleSizeOption[] | null
  sampleSize: number | null
  endMeasurements: {
    isComplete: null | boolean
    pvalue: null | number
  }
}

export interface Round {
  name?: string
  randomSeed?: string
  riskLimit?: number
  contests: RoundContest[]
  startedAt: string
  endedAt: string | null
  jurisdictions?: {
    [key: string]: {
      numBallots: number
    }
  }
}

export interface Audit {
  name: string
  riskLimit: string
  randomSeed: string
  contests: Contest[]
  jurisdictions: Jurisdiction[]
  rounds: Round[]
}
