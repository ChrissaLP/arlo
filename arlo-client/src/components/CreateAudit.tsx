import React, { useState } from 'react'
import styled from 'styled-components'
import { toast } from 'react-toastify'
import { RouteComponentProps, Link } from 'react-router-dom'
import FormButton from './Form/FormButton'
import { api, checkAndToast } from './utilities'
import { ICreateAuditParams, IErrorResponse } from '../types'
import { useAuthDataContext } from './UserContext'

const Button = styled(FormButton)`
  margin: 30px 0 0 0;
`

const Wrapper = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  @media (min-width: 500px) {
    width: 400px;
  }
`

const AuditLink = styled(Link)`
  display: block;
  margin: 10px;

  &:first-of-type {
    margin-top: 30px;
  }
`

const CreateAudit = ({ history }: RouteComponentProps<ICreateAuditParams>) => {
  const { isAuthenticated, meta } = useAuthDataContext()

  const [loading, setLoading] = useState(false)
  const onClick = async () => {
    try {
      setLoading(true)
      const data = isAuthenticated
        ? // eslint-disable-next-line @typescript-eslint/camelcase
          { organization_id: meta!.organizations[0].id }
        : {}
      const response: { electionId: string } | IErrorResponse = await api(
        '/election/new',
        {
          method: 'POST',
          body: JSON.stringify(data),
        }
      )
      if (checkAndToast(response)) {
        return
      }
      const { electionId } = response
      history.push(`/election/${electionId}`)
    } catch (err) {
      toast.error(err.message)
    }
  }
  return (
    <Wrapper>
      <img height="50px" src="/arlo.png" alt="Arlo, by VotingWorks" />
      <Button
        type="button"
        intent="primary"
        fill
        large
        onClick={onClick}
        loading={loading}
        disabled={loading}
      >
        Create a New Audit
      </Button>
      {!isAuthenticated ? (
        <>
          <Button
            type="button"
            intent="primary"
            fill
            large
            onClick={
              /* istanbul ignore next */
              () => window.location.replace('/auth/auditadmin/start')
            }
            loading={loading}
            disabled={loading}
          >
            Log in as an Audit Admin
          </Button>
          <Button
            type="button"
            intent="primary"
            fill
            large
            onClick={
              /* istanbul ignore next */
              () => window.location.replace('/auth/jurisdictionadmin/start')
            }
            loading={loading}
            disabled={loading}
          >
            Log in as a Jurisdiction Admin
          </Button>
        </>
      ) : (
        <>
          {meta!.organizations.length > 0 &&
            meta!.organizations.map(o =>
              o.elections.map(election => (
                <AuditLink
                  to={`/election/${election.id}`}
                  key={election.id}
                  className="bp3-button bp3-intent-primary"
                >
                  {election.name || 'Not named yet'}
                  {election.state && ` (${election.state})`}
                </AuditLink>
              ))
            )}
          {meta!.jurisdictions.length > 0 &&
            meta!.jurisdictions.map(({ election }) => (
              <AuditLink
                to={`/election/${election.id}`}
                key={election.id}
                className="bp3-button bp3-intent-primary"
              >
                {election.name || 'Not named yet'}
                {election.state && ` (${election.state})`}
              </AuditLink>
            ))}
        </>
      )}
    </Wrapper>
  )
}

export default CreateAudit
