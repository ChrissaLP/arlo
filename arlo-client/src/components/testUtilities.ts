import React from 'react'
import { RenderResult, act, render } from '@testing-library/react'
import { createLocation, createMemoryHistory } from 'history'
import { match as routerMatch } from 'react-router-dom'
import { IAudit, IUserMeta, IRound, IAuditSettings, IBallot } from '../types'
import { IJurisdictions } from './Audit/Setup/useParticipantsApi'
import { IContests } from './Audit/Setup/Contests/types'

type MatchParameter<Params> = { [K in keyof Params]?: string }

const generateUrl = <Params extends MatchParameter<Params>>(
  path: string,
  params: Params
): string => {
  let tempPath = path

  for (const param in params) {
    /* istanbul ignore else */
    if (Object.prototype.hasOwnProperty.call(params, param)) {
      const value = params[param]
      tempPath = tempPath.replace(`:${param}`, value as NonNullable<
        typeof value
      >)
    }
  }

  return tempPath
}

export const routerTestProps = <Params extends MatchParameter<Params> = {}>(
  path: string,
  params: Params
) => {
  const match: routerMatch<Params> = {
    isExact: false,
    path,
    url: generateUrl(path, params),
    params,
  }
  const history = createMemoryHistory()
  const location = createLocation(match.url)

  return { history, location, match }
}

/** Credit to https://stackoverflow.com/a/56452779 for solution to mocking React Router props */

export const regexpEscape = (s: string) => {
  /* eslint-disable no-useless-escape */
  return s.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')
}

export const asyncActRender = async (
  component: React.ReactElement
): Promise<RenderResult> => {
  let result: RenderResult
  await act(async () => {
    result = render(component)
  })
  return result!
}

export const generateApiMock = ({
  statusReturn,
  authReturn,
  roundReturn,
  jurisdictionReturn,
  settingsReturn,
  ballotsReturn,
  contestsReturn,
}: {
  statusReturn?: IAudit | Error | { status: 'ok' }
  authReturn?: IUserMeta | Error
  roundReturn?: { rounds: IRound[] } | Error | { status: 'ok' }
  jurisdictionReturn?:
    | { jurisdictions: IJurisdictions }
    | Error
    | { status: 'ok' }
  settingsReturn?: IAuditSettings | Error
  ballotsReturn?: { ballots: IBallot[] } | Error
  contestsReturn?: IContests | Error
}) => async (
  endpoint: string
): Promise<
  | IAudit
  | IUserMeta
  | { rounds: IRound[] }
  | { jurisdictions: IJurisdictions }
  | IAuditSettings
  | { ballots: IBallot[] }
  | IContests
  | Error
  | { status: 'ok' }
> => {
  if (endpoint === '/election/1/audit/status' && statusReturn)
    return statusReturn
  if (endpoint === '/auth/me' && authReturn) return authReturn
  if (endpoint === '/election/1/round' && roundReturn) return roundReturn
  if (endpoint === '/election/1/jurisdiction' && jurisdictionReturn)
    return jurisdictionReturn
  if (endpoint === '/election/1/settings' && settingsReturn)
    return settingsReturn
  if (
    endpoint.match(
      /\/election\/[^/]+\/jurisdiction\/[^/]+\/round\/[^/]+\/ballot-list/
    ) &&
    ballotsReturn
  )
    return ballotsReturn
  if (endpoint === '/election/1/contest' && contestsReturn)
    return contestsReturn
  return new Error(`missing mock for ${endpoint}`)
}

export default {
  routerTestProps,
}
