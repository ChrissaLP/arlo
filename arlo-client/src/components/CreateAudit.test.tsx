import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import { toast } from 'react-toastify'
import { RouteComponentProps } from 'react-router-dom'
import CreateAudit from './CreateAudit'
import { ICreateAuditParams } from '../types'
import { routerTestProps } from './testUtilities'
import { api, checkAndToast } from './utilities'

const apiMock = api as jest.Mock<ReturnType<typeof api>, Parameters<typeof api>>
const checkAndToastMock = checkAndToast as jest.Mock<
  ReturnType<typeof checkAndToast>,
  Parameters<typeof checkAndToast>
>

jest.mock('./utilities')

checkAndToastMock.mockImplementation(() => false)

const routeProps: RouteComponentProps<ICreateAuditParams> = routerTestProps(
  '/election/:electionId',
  {
    electionId: '1',
  }
)

const toastSpy = jest.spyOn(toast, 'error').mockImplementation()

const historySpy = jest.spyOn(routeProps.history, 'push').mockImplementation()

afterEach(() => {
  apiMock.mockClear()
  checkAndToastMock.mockClear()
  toastSpy.mockClear()
  historySpy.mockClear()
})

describe('CreateAudit', () => {
  it('renders correctly', () => {
    const { container } = render(<CreateAudit {...routeProps} />)
    expect(container).toMatchSnapshot()
  })

  it('calls the /election/new endpoint', async () => {
    apiMock.mockImplementation(async () => ({ electionId: '1' }))
    const { getByText } = render(<CreateAudit {...routeProps} />)

    fireEvent.click(getByText('Create a New Audit'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toBeCalledTimes(1)
      expect(apiMock.mock.calls[0][0]).toBe('/election/new')
      expect(historySpy).toBeCalledTimes(1)
      expect(historySpy.mock.calls[0][0]).toBe('/election/1')
    })
  })

  it('handles error responses from server', async () => {
    apiMock.mockImplementation(async () => ({ electionId: '1' }))
    checkAndToastMock.mockImplementation(() => true)
    const { getByText } = render(<CreateAudit {...routeProps} />)

    fireEvent.click(getByText('Create a New Audit'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toBeCalledTimes(1)
      expect(apiMock.mock.calls[0][0]).toBe('/election/new')
      expect(checkAndToastMock).toBeCalledTimes(1)
      expect(historySpy).toBeCalledTimes(0)
      expect(toastSpy).toBeCalledTimes(0)
    })
  })

  it('handles 404 responses from server', async () => {
    apiMock.mockImplementation(async () => {
      throw new Error('404')
    })
    checkAndToastMock.mockImplementation(() => true)
    const { getByText } = render(<CreateAudit {...routeProps} />)

    fireEvent.click(getByText('Create a New Audit'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toBeCalledTimes(1)
      expect(apiMock.mock.calls[0][0]).toBe('/election/new')
      expect(checkAndToastMock).toBeCalledTimes(0)
      expect(historySpy).toBeCalledTimes(0)
      expect(toastSpy).toBeCalledTimes(1)
    })
  })
})
