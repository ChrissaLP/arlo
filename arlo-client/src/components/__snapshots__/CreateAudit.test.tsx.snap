// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAudit lists associated elections for authenticated AA user 1`] = `
<div>
  <div
    class="sc-htpNat dgtpPg"
  >
    <img
      alt="Arlo, by VotingWorks"
      height="50px"
      src="/arlo.png"
    />
    <button
      class="bp3-button bp3-fill bp3-large bp3-intent-primary sc-bwzfXH fdlCQb sc-bdVaJa gLVjYw"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        Create a New Audit
      </span>
    </button>
    <a
      class="bp3-button bp3-intent-primary sc-bxivhb cQgiba"
      href="/election/election-1"
    >
      Not named yet
       (NY)
    </a>
    <a
      class="bp3-button bp3-intent-primary sc-bxivhb cQgiba"
      href="/election/election-2"
    >
      Election Two
       (FL)
    </a>
    <a
      class="bp3-button bp3-intent-primary sc-bxivhb cQgiba"
      href="/election/election-3"
    >
      Election Three
      
    </a>
    <a
      class="bp3-button bp3-intent-primary sc-bxivhb cQgiba"
      href="/election/election-4"
    >
      Election Four
       (WA)
    </a>
  </div>
</div>
`;

exports[`CreateAudit lists associated elections for authenticated JA user 1`] = `
<div>
  <div
    class="sc-htpNat dgtpPg"
  >
    <img
      alt="Arlo, by VotingWorks"
      height="50px"
      src="/arlo.png"
    />
    <button
      class="bp3-button bp3-fill bp3-large bp3-intent-primary sc-bwzfXH fdlCQb sc-bdVaJa gLVjYw"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        Create a New Audit
      </span>
    </button>
    <a
      class="bp3-button bp3-intent-primary sc-bxivhb cQgiba"
      href="/election/election-1"
    >
      Election One
       (NY)
    </a>
    <a
      class="bp3-button bp3-intent-primary sc-bxivhb cQgiba"
      href="/election/election-2"
    >
      Not named yet
      
    </a>
  </div>
</div>
`;

exports[`CreateAudit renders correctly 1`] = `
<div>
  <div
    class="sc-htpNat dgtpPg"
  >
    <img
      alt="Arlo, by VotingWorks"
      height="50px"
      src="/arlo.png"
    />
    <button
      class="bp3-button bp3-fill bp3-large bp3-intent-primary sc-bwzfXH fdlCQb sc-bdVaJa gLVjYw"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        Create a New Audit
      </span>
    </button>
    <button
      class="bp3-button bp3-fill bp3-large bp3-intent-primary sc-bwzfXH fdlCQb sc-bdVaJa gLVjYw"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        Log in as an Audit Admin
      </span>
    </button>
    <button
      class="bp3-button bp3-fill bp3-large bp3-intent-primary sc-bwzfXH fdlCQb sc-bdVaJa gLVjYw"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        Log in as a Jurisdiction Admin
      </span>
    </button>
  </div>
</div>
`;
